hal/rfcs/0000-spi-overhaul.md

```diff
@@ -0,0 +1,254 @@
+| `spi_transfer_async(..., uint32_t event, ...)` | argument removed | the callback will now be invoked on any event with the event as an argument.
+| `spi_busy(spi)` | removed | This method is never used outside of some of the hal implementations.
+| `spi_master_write(...)` `spi_slave_write(...)` `spi_slave_read(...)` `spi_slave_receive(...)` | removed | Writing data symbol by symbol is very inefficient and should not be encouraged. It can still be achieved by passing a pointer to the symbol to `spi_transfer`.
+| `spi_active(spit_t *obj)` | removed | as the async call back is now always invoked on async operation termination (unless cancelled), this status can be tracked from driver layer without any hal request.
+| `uint32_t spi_irq_handler_asynch(spi_t *obj)` | removed | as the event is now passed as an argument to the callback this method is no longer required.
+
+### The new API
+
+```c
+typedef struct {
+    /** Minimum frequency supported must be set by target device and it will be assessed during
+     *  testing.
+     */
+    uint32_t    minimum_frequency;
+    /** Maximum frequency supported must be set by target device and it will be assessed during
+     *  testing.
+     */
+    uint32_t    maximum_frequency;
+    /** Each bit represents the corresponding word length. lsb => 1bit, msb => 32bits.
+     *
+     * For example, if the peripheral supports 8 bits, 12bits to 16bits and 32bits,
+     * the value should be 0x8000F880.
+     */
+    uint32_t    word_length;
+    bool        support_slave_mode; /**< If true, the device can handle SPI slave mode using hardware management on the specified ssel pin. */
+    bool        half_duplex; /**< If true, the device also supports SPI transmissions using only 3 wires. */
+} spi_capabilities_t;
+
+typedef enum _spi_mode_t {
+  SPI_MODE_IDLE_LOW_SAMPLE_FIRST_EDGE,
+  SPI_MODE_IDLE_LOW_SAMPLE_SECOND_EDGE,
+  SPI_MODE_IDLE_HIGH_SAMPLE_FIRST_EDGE,
+  SPI_MODE_IDLE_HIGH_SAMPLE_SECOND_EDGE,
+} spi_mode_t;
+
+typedef enum _spi_bit_ordering_t {
+  SPI_BIT_ORDERING_MSB_FIRST,
+  SPI_BIT_ORDERING_LSB_FIRST,
+} spi_bit_ordering_t;
+
+typedef void (*spi_async_handler_f)(spi_t *obj, void *ctx, spi_async_event_t event);
+
+/**
+ * Returns a variant of the SPIName enum uniquely identifying a SPI peripheral of the device.
+ */
+SPIName spi_get_module(PinName mosi, PinName miso, PinName mclk);
+/**
+ * Fills the given spi_capabilities_t structure with the capabilities of the given peripheral.
+ */
+void spi_get_capabilities(SPIName name, PinName ssel, spi_capabilities_t *cap);
+
+void spi_init(spi_t *obj, bool is_slave, PinName mosi, PinName miso, PinName mclk, PinName ssel);
+void spi_format(spi_t *obj, uint8_t bits, spi_mode_t mode, spi_bit_ordering_t bit_ordering);
+uint32_t spi_frequency(spi_t *obj, uint32_t hz);
+void spi_transfer(spi_t *obj, const void *tx, uint32_t tx_len, void *rx, uint32_t rx_len, const void *fill_symbol);
```

Comment:
`spi_transfer()` shell return the number of symbol clocked on the bus during this transfer.

This should be added to the defined behavior section also.